python_utils.py

```diff
@@ -0,0 +1,376 @@
+# coding: utf-8
+#
+# Copyright 2019 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Feature detection utilities for Python 2 and Python 3."""
+
+import os
+import sys
+
+_FUTURE_PATH = os.path.join('third_party', 'future-0.17.1')
+sys.path.insert(0, _FUTURE_PATH)
+
+# pylint: disable=wrong-import-position
+# pylint: disable=wrong-import-order
+import builtins  # isort:skip
+import future.utils  # isort:skip
+import past.builtins  # isort:skip
+import past.utils  # isort:skip
+# pylint: enable=wrong-import-order
+# pylint: enable=wrong-import-position
+
+
+BASESTRING = past.builtins.basestring
+INPUT = builtins.input
+MAP = builtins.map
+NEXT = builtins.next
+OBJECT = builtins.object
+RANGE = builtins.range
+ROUND = builtins.round
+STR = builtins.str
+ZIP = builtins.zip
+
+
+def string_io(buffer_value=''):
+    """Returns StringIO from StringIO module if run under Python 2 and from io
+    module if run under Python 3.
+
+    Returns:
+        StringIO.StringIO or io.StringIO. The StringIO object.
+    """
+    try:
+        from StringIO import StringIO  # pylint: disable=import-only-modules
+    except ImportError:
+        from io import StringIO  # pylint: disable=import-only-modules
+    return StringIO(buffer_value)
+
+
+def get_args_of_function(function_node, args_to_ignore):
+    """Extracts the arguments from a function definition.
+
+    Args:
+        function_node: ast.FunctionDef. Represents a function.
+        args_to_ignore: list(str). Ignore these arguments in a function
+        definition.
+
+    Returns:
+        list(str). The args for a function as listed in the function
+        definition.
+    """
+    try:
+        return [
+            a.arg for a in function_node.args.args if a.arg not in
+            args_to_ignore]
+    except AttributeError:
+        return [
+            a.id for a in function_node.args.args if a.id not in args_to_ignore]
+
+
+def open_file(filename, mode, encoding='utf-8'):
+    """Open file and return a corresponding file object.
+
+    Args:
+        filename: str. The file to be opened.
+        mode: str. Mode in which the file is opened.
+        encoding: str. Encoding in which the file is opened.
+
+    Returns:
+        _io.TextIOWrapper. The file object.
+    """
+    import io
+    try:
+        return io.open(filename, mode, encoding=encoding)
+    except:
+        raise IOError('Unable to open file: %s' % filename)
```

Comment:
I don't think this is necessary?

Reply:
I just used a catch so that we will get the same error with python 2 and 3. Because python 3 will throw a FileNotFoundError while python 2 will throw an IOError.

Reply:
FileNotFoundError is a subclass of IOError, so it shouldn't be a breaking change. Should be fine to remove this.

Reply:
Ah nvm, it doesn't exist in Python 3. This fine is change, we'll just remember to avoid this after migration.